aten/src/ATen/native/mkldnn/Normalization.cpp

```diff
@@ -39,38 +53,86 @@ std::tuple<Tensor, Tensor, Tensor> mkldnn_batch_norm(
-    // ideep::tensor saved_var;
-    // ideep::batch_normalization_forward_training::compute<AllocForMKLDNN>(
-    //     x, w, b, y, saved_mean, saved_var, m, v, momentum, eps);
-    // return std::make_tuple(
-    //     new_with_itensor_mkldnn(std::move(y), input.options()),
-    //     new_with_itensor_mkldnn(std::move(saved_mean), input.options()),
-    //     new_with_itensor_mkldnn(std::move(saved_var), input.options()));
+    ideep::tensor saved_mean;
+    ideep::tensor saved_var;
+    if (use_running_stat) {
+      ideep::tensor m = itensor_view_from_dense(running_mean);
+      ideep::tensor v = itensor_view_from_dense(running_var);
+      ideep::batch_normalization_forward_training::compute<AllocForMKLDNN>(
+          x, w, b, y, saved_mean, saved_var, m, v, momentum, eps);
+    } else {
+      ideep::batch_normalization_forward_training::compute<AllocForMKLDNN>(
+          x, w, b, y, saved_mean, saved_var, momentum, eps);
+    }
+    return std::make_tuple(
+        new_with_itensor_mkldnn(std::move(y), input.options()),
+        new_with_itensor_mkldnn(std::move(saved_mean), input.options()),
+        new_with_itensor_mkldnn(std::move(saved_var), input.options()));
   } else {
-    AT_ASSERTM(input.dim() == 4 || input.dim() == 5,
-               "mkldnn_batch_norm: currently mkldnn only support 2d and 3d batchnorm");
-    ideep::batch_normalization_forward_inference::compute<AllocForMKLDNN>(
-        x, m, v, w, b, y, eps);
+    if (use_running_stat) {
+      ideep::tensor m = get_mkldnn_tensor(running_mean);
+      ideep::tensor v = get_mkldnn_tensor(running_var);
+      ideep::batch_normalization_forward_inference::compute<AllocForMKLDNN>(
+          x, m, v, w, b, y, eps);
+    } else {
+      ideep::batch_normalization_forward_inference::compute<AllocForMKLDNN>(
+          x, w, b, y, eps);
+    }
     return std::make_tuple(
         new_with_itensor_mkldnn(std::move(y), input.options()),
         new_with_itensor_mkldnn(ideep::tensor{}, input.options()),
         new_with_itensor_mkldnn(ideep::tensor{}, input.options()));
   }
 }
 
+std::tuple<Tensor, Tensor, Tensor> mkldnn_batch_norm_backward(const Tensor& grad_output,
+    const Tensor& input,
+    const Tensor& weight,
+    const Tensor& running_mean,
+    const Tensor& running_var,
+    const Tensor& save_mean,
+    const Tensor& save_invstd,
+    bool train,
+    double eps,
+    std::array<bool,3> grad_input_mask) {
+  AT_ASSERTM(train, "mkldnn_batch_norm_backward: currently mkldnn only support train model");
+
+  ideep::tensor& grady = itensor_from_mkldnn(grad_output);
+  ideep::tensor& x = itensor_from_mkldnn(input);
+  ideep::tensor w = itensor_view_from_dense(weight);
+  ideep::tensor& m = itensor_from_mkldnn(save_mean);
+  ideep::tensor& v = itensor_from_mkldnn(save_invstd);
+
+  ideep::tensor gradx, gradw, gradb;
+  ideep::batch_normalization_backward::compute<AllocForMKLDNN>(
+      x, m, v, grady, w, gradx, gradw, gradb, eps);
+
+  if (weight.is_mkldnn()) {
```

Comment:
Ok if weigh could be is_mkldnn, how line 115 supposed to work `ideep::tensor w = itensor_view_from_dense(weight);`